### PR TITLE
feat: add `CYPRESS_SKIP_VERIFY` option to suppress verification

### LINKED ISSUE #22243: Being able to disable cypress verify step
Hello!

### What would you like?

I'd like to be able to disable the verification process (`cypress verify`) that is triggered on `cypress run`.

### Why is this needed?

Our app that is running in Docker/Jenkins suffers of the timeout issue that was raised a few times (#4624 #6082)

We recently increased the timeout to 60s thanks to the new [`CYPRESS_VERIFY_TIMEOUT`](https://github.com/cypress-io/cypress/blob/576519e465af28e2ed3f0b2c181fbe8f17e348a2/cli/lib/tasks/verify.js#L18) env variable; but the verification still timeouts from time to time.

Something that is not clear to me is the rationale behind this verification process 👇 

- Running it in a local environment as a "defensive" mechanism definitely makes sense, as:
  - Cypress may run on a broad range of configurations, and has specific requirements
  - The host configuration may have changed between two runs
  - The output is helpful to fix the host configuration if needed
- Running it in a containerized environment is debatable as, after said env has been properly configured once, subsequent builds are supposed to be identical
  - And in the case of a CI misconfiguration, I would also prefer the tests to start running and fail immediately, than waiting for the timeout to happen

Thank you for reading! 🌲

### PR DESCRIPTION
**Pull request by joshuajtward**
- Closes #22243

### Additional details
The PR adds a flag `CYPRESS_SKIP_VERIFY` to prevent the verification steps from running. These steps are semi-redundant in CI environments, and regularly timeout leading to false negatives in test suites. 

### Steps to test
- Delete the Cypress cache on your machine and run Cypress with `CYPRESS_SKIP_VERIFY=true` in your env
- Confirm that the verification messages are not shown ("Looks like this is your first time running Cypress N.N.N...")

### How has the user experience changed?
If a user opts in they will not get the verification on Cypress startup

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: cli/CHANGELOG.md [hunk 1]
```diff
@@ -16,6 +16,10 @@ _Released 7/16/2024_
 - Replaced `json-lint` with `json-parse-even-better-errors`. This removes the [CVE-2021-23358](https://nvd.nist.gov/vuln/detail/CVE-2021-23358) vulnerability being reported in security scans. Addresses [#28207](https://github.com/cypress-io/cypress/issues/28207).
 - Updated `minimatch` from `3.0.4` to `3.1.2`. Addressed in [#29821](https://github.com/cypress-io/cypress/pull/29821).
 
+**Misc:**
+
+- Added `CYPRESS_NO_VERIFY` flag to enable suppression of Cypress verification checks. Addresses [#22243](https://github.com/cypress-io/cypress/issues/22243)
```
Comment (MikeMcC399) [inline cli/CHANGELOG.md]: `CYPRESS_NO_VERIFY` is not the name of the environment variable in the code, which is instead `CYPRESS_SKIP_VERIFY`

**Review by MikeMcC399 — CHANGES_REQUESTED** [review]
changelog does not match code

**Comment by MikeMcC399** [comment]
@joshuajtward 

> ### Additional details
> The PR adds a flag `CYPRESS_SKIP_VERIFY` to prevent the verification steps from running. These steps are semi-redundant in CI environments, and regularly timeout leading to false negatives in test suites.

Did you open an issue about timeouts so this could be investigated? I'm seeing some other problems in CI environments with `cypress verify` where the check is uncovering some other issues, so I wonder if it is a good idea to suppress this check.

**Comment by joshuajtward** [comment]
Hi @MikeMcC399! I have not opened an issue about the timeouts, but there has been a lot of discussion in this repo about this issue already (see for example #6082). It has been partially addressed by adding the ability to set a long timeout, but I would like to add the option to suppress the error so that we can:

1. reduce the false negatives in our CI pipelines
2. see if suppressing it raises any more informative logs or errors

**Comment by MikeMcC399** [comment]
@joshuajtward 

- Thank you very much for the reference to issue #6082 ! I didn't know about that. I haven't seen this happen recently with Cypress Docker images, but maybe this depends on the environment?

**Comment by joshuajtward** [comment]
Yes @MikeMcC399, I'm pretty sure something is going awry with the child tasks which interrogate the installed vs expected binaries, but it's pretty hard to narrow it down because it only happens on around 2% of jobs and usually passes on a re-run. I don't think it's a sensible use of resources to put `DEBUG=cypress:*` in all our CI runs...

For our use cases we don't much care about these integrity checks as we're pretty confident in our CI environment's stability (I think that will be a commonly held sentiment),  our jobs are all running on the same container infrastructure, and using the same caches. We don't want to have to implement a retry shell script or other such workarounds, so using a flag to opt out seems like a better way forward for us.

If our use of this flag helps to highlight a root cause or underlying problem I will be happy to raise it as an issue in this repo!

**Comment by MikeMcC399** [comment]
@joshuajtward 

Thanks again for your explanation and background! Which CI provider are you using and which package manager (npm, Yarn or pnpm)?

**Comment by joshuajtward** [comment]
@MikeMcC399 we're using GitHub Actions with a mix of Linux, Windows, and MacOS, and a mix of `yarn` and `npm` across a number of repos

**Comment by MikeMcC399** [comment]
The new environment variable would need to be documented in the table https://docs.cypress.io/guides/references/advanced-installation#Environment-variables so this would mean that a release branch for `13.14.0` (or later if delayed) would need to be created in the docs repo to prepare for release.

image

**Review by MikeMcC399 — CHANGES_REQUESTED** [review]
The documentation aspect needs to be covered.

**Comment by joshuajtward** [comment]
@MikeMcC399 docs PR is here: https://github.com/cypress-io/cypress-documentation/pull/5889

**Comment by joshuajtward** [comment]
@MikeMcC399 @jennifer-shehane is there anything else I can help with to get this moving?

**Comment by MikeMcC399** [comment]
@joshuajtward 

> is there anything else I can help with to get this moving?

Yes, you will need to resolve the conflicts in the branch. Unfortunately this can happen as changes are committed to the `develop` branch.

This is mentioned in the submitter's guide https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#pull-requests

**Comment by joshuajtward** [comment]
@MikeMcC399 sorry I hadn't seen there were more conflicts. I have updated that now, is there anything else?

**Comment by MikeMcC399** [comment]
@joshuajtward 

> sorry I hadn't seen there were more conflicts. I have updated that now

👍🏻 

> is there anything else?

CI has one failing test, however that may be CircleCI flake. I expect that @jennifer-shehane will be able to provide more concrete feedback as the Cypress.io team is more familiar with the individual tests than I am. They can also re-run individual tests. 

Since I'm only an external contributor, I don't have write privileges to CircleCI, to the repo or to the PR approval process. I can just provide comments.

**Comment by jennifer-shehane** [comment]
@joshuajtward Just awaiting @AtofStryker review at this point

File: cli/lib/tasks/verify.js [hunk 2]
```diff
@@ -258,8 +258,13 @@ const start = (options = {}) => {
     force: false,
     welcomeMessage: true,
     smokeTestTimeout: VERIFY_TEST_RUNNER_TIMEOUT_MS,
+    skipVerify: util.getEnv('CYPRESS_SKIP_VERIFY') || 'false',
```
Comment (AtofStryker) [inline cli/lib/tasks/verify.js]: Can we get this to just work with booleans?
```suggestion
    skipVerify: util.getEnv('CYPRESS_SKIP_VERIFY') === 'true',
```

File: cli/lib/tasks/verify.js [hunk 3]
```diff
@@ -258,8 +258,13 @@ const start = (options = {}) => {
...
   })
 
+  if (options.skipVerify === 'true') {
```
Comment (AtofStryker) [inline cli/lib/tasks/verify.js]:
```suggestion
  if (options.skipVerify) {
```

File: cli/lib/tasks/verify.js [hunk 4]
```diff
@@ -258,8 +258,13 @@ const start = (options = {}) => {
...
+    return Promise.resolve()
```
Comment (AtofStryker) [inline cli/lib/tasks/verify.js:267]:
```suggestion
    debug('skipping verification of the Cypress app')
    return Promise.resolve()
```

**Review by AtofStryker — CHANGES_REQUESTED** [review]

**Comment by joshuajtward** [comment]
Good shout @AtofStryker, that reads more nicely - I've updated it now

**Comment by jennifer-shehane** [comment]
@joshuajtward There's a linting error 
Screenshot 2024-08-09 at 10 25 04 AM

**Comment by joshuajtward** [comment]
@AtofStryker I have fixed this up and it should be ready for a re-review, thanks!

**Comment by AtofStryker** [comment]
@joshuajtward Thank you for getting the changes addressed! One last question on my end. I created a binary of your code changes on https://github.com/cypress-io/cypress/commit/add6f09401f27489d2808ba4bc87096fd2dd42bf. Are you able to download the binary from the link and verify the changes work as expected?

**Comment by joshuajtward** [comment]
> @joshuajtward Thank you for getting the changes addressed! One last question on my end. I created a binary of your code changes on [add6f09](https://github.com/cypress-io/cypress/commit/add6f09401f27489d2808ba4bc87096fd2dd42bf). Are you able to download the binary from the link and verify the changes work as expected?

Hi @AtofStryker, I have installed the `linux arm64` binary and run a few tests, it appears to work as expected 

Without the flag it works as normal: 

No flag

With the flag the verify steps are skipped:

With flag

And if the binary is missing then it fails with an illustrative error: 

No binary

Please let me know if there's any other testing you would like me to do!

**Comment by MikeMcC399** [comment]
@joshuajtward 

I notice that you're running quite an old version of Yarn (`v1.18.0`) from 2019. Yarn `v1.22.22` is the current stable version (see https://classic.yarnpkg.com/en/docs/install).

- Edit: See related https://github.com/cypress-io/cypress/pull/30048, but probably not an issue for this PR.

**Review by AtofStryker — APPROVED** [review]

**Comment by joshuajtward** [comment]
@jennifer-shehane are we all set to merge? 🙏

**Review by cacieprins — APPROVED** [review]

**Pull request merged**